### ISSUE TITLE
Invalid JSON RPC Error

### ISSUE THREAD

**Issue by roschler** [trufflesuite/ganache#1057]
_PREFACE: this may be my fault because I am running a custom version of Ganache that filters out **eth_blockNumber** calls from the log output.  But the error dialogue Ganache put up asked me to raise a GitHub issue so I am.  When I did the same operation on the the "stock" (unmodified) version of Ganache I ran first, I did not get this error._

I had just added my project (first and only project) to my Quickstart workspace. Note, I had the Chrome debugger open when running Ganache.  I mention this in case it has something to do with the problem occurring.  I have noticed that electron-forge apps appear to put a heavy load on the debugger. 

PLATFORM: linux
GANACHE VERSION: 2.0.0-beta.1

EXCEPTION:
```
Not connected to child process
```

**Comment by seesemichaelj** [comment]
@roschler Thanks for reporting this regardless! Do you always get this error or just sometimes? And if sometimes, does it happen frequently? Do you have a link to the forked version of Ganache that I can try to reproduce?

Thanks!

**Comment by roschler** [comment]
@seesemichaelj Hi Mike.  I created a private repo for reproducing the error.  I have just sent you an invite to the repo.  The repo was designed to reproduce the same chain of different Ethereum transactions I execute during a game in my dApp.  In my full dApp, the error occurs reliably on the very last step of the game.

However, I discovered when during the creation of the bug repo that in that code I get the error reliably the instant I execute a Web3 `getAccounts()` call, every time.  The repo has full directions on how to execute the test and other details in its README.  Note, the error happens with `ganache-cli` and the latest version of Ganache GUI, so my belief is that it's a problem with ganache-core.

If you have any questions let me know and thanks in advance for taking a look at this.

**Comment by seesemichaelj** [comment]
After some further investigation and troubleshooting over Gitter with @roschler, we got to a better configuration by cleaning `node_modules`. Now waiting to see if the issue still stands

**Comment by roschler** [comment]
I am finally able to execute my last transaction in my dApp.  The overall solution came down to the disabling of the socket keep-alive check and the need to download an exact, specific version of the `Web3` JS package.  I'll update the GitHub issue with this information.  A giant thanks @seesemichaelj Mike for all your help and for your hard work on the Ganache.  Ganache chocolate is an **indispensable** tool for Ethereum developers.

I am putting a list of useful GitHub issue links here for others that may encounter the more difficult problems I had recently:

### Related to Web3 package problems.  (Note, this is time sensitive information and likely to change):

https://github.com/ethereum/web3.js/issues/966

Solution is in this message:

https://github.com/ethereum/web3.js/issues/966#issuecomment-451133030

### Related to keep-alive timeouts that cause the connection to Ganache to be dropped by Ganache core (this thread):

https://github.com/trufflesuite/ganache/issues/1057#issuecomment-451991759

### Problems related to values that should be BigNumber values:

SUGGESTION: **Make all Wei related values you pass to Ethereum calls like `sendSignedTransaction()` etc. BigNumber values or you will suffer a great deal with intermittent _seemingly_ unrelated errors like "Invalid JSON RPC response"!**

### Problems related to use of the wrong Buffer object format

**TL;DR version**

Always create Buffer objects with `ethereum_util.toBuffer(string)` and make sure the string passed to that method is prepended with the '0x' hex format specifier.

**Full version**

You can find posts on the web indicating you should use `Buffer.from(string)` or the deprecated form `let buff = new Buffer(string)` to convert a private key in string format to a buffer. But that's incompatible with whatever buffer format packages like those in the `ethereumjs` suite are using. 

Buffer objects created with Buffer.from(string) from a private key are 64 in length. Buffer objects created by the dedicated method `ethereum_util.toBuffer(string)` using a private key are exactly half that length; they are 32 in length. I traced through the `ethereumjs-tx` code and the `sign()` method throws an error if the Buffer object is not _exactly 32 in length_.  This makes sense since that is what their library `toBuffer()` method produces. So using Buffer.from(string) is just wrong if you are using the `ethereumjs-tx` package(s) at the very least, probably in the general case as well when working with Ethereum transactions..

However, when using `ethereum_util.toBuffer(string)` you **must** put 0x in front of it or you get the same double-length Buffer object that `Buffer.from(string)` produces. It seems that the code in `ethereumjs-tx` knows what to do with the hex format specified while the generic Node.JS code in `Buffer.from(string)` does not. If you try putting '0x' in front of a string when calling Buffer.from(string) you still get a Buffer object that is still 64 in length.

NOTE to Truffle Web3 team: perhaps all the calls that involve numeric values should have a Prologue that throws an error if _any_ of the parameters are not in BigNumber format?  I'd suggest auto-convert to BigNumber but that could create the kind of "hidden" errors that create more problems than solutions.

I hope this helps others.

**Comment by davidmurdoch** [comment]
@roschler on the Buffer issue, you can still use node's `Buffer.from` with a privatekey string but you need to specify the encoding of the data (https://nodejs.org/api/buffer.html#buffer_class_method_buffer_from_string_encoding).

`Buffer.from("1234")` will encode the _utf8_ string "1234" as `[31, 32, 33, 34]` whereas `Buffer.from("1234", "hex")` will encode the hex data as `[12, 34]`.

By prepending `0x` to the ethereumjs-utils' `toBuffer` method you are instructing it to treat the string as hex-encoded data.

**Comment by roschler** [comment]
Good point @davidmurdoch David, thanks.  These are the variations that can really drive you mad during a late night coding/debuggin session.  I'm referring to the fact that one library does accept and handle properly the "0x" format specified and the other requires you to specify the format explicitly as a second parameter with the value "hex" as you point out in your post.

Typically variety is the spice of life but here it is more akin to the spice that ends up giving you gas, and I don't mean the good kind like in the context of an Ethereum transaction! :)
